### DISCUSSION TITLE
COPY and nextval('seq')

### DISCUSSION THREAD

**Discussion by greyltc** [psycopg/psycopg#407] [Q&A] [answered]
I've read the docs on COPY: https://www.psycopg.org/psycopg3/docs/basic/copy.html and the basic usage outlined there is working just fine for me, but I'd like to copy into a table using a sequence that I've made with `CREATE TEMPORARY SEQUENCE seq` to fill the values for one column (and copy values for others). It seems the only chance that this can work is if I can somehow get the copy to work with `nextval('seq')`, since setting up DEFAULT with a temporary sequence is a horrible idea.

So is there some way to use a function like `nextval()` with COPY ... FROM STDIN?

**Comment by dvarrazzo** [comment, marked as answer]
As far as I know, COPY protocol only take data, not expressions, so I think you need to have nextval in the default of a field. You can hear from pgsql-general if there's a different way, and if there is, psycopg should support it

**Comment by greyltc** [comment]
That's what I expected and what I was afraid of.
Unfortunately it's not realistically possible to use nextval on a temporary sequence as DEFAULT for a column either. So all taken together, this means that using a temporary sequence in a table that one wants to `COPY ... FROM ...` into is essentially not possible. I'll have to generate the sequence numbers myself programmatically in python and INSERT/COPY them without postgres being able to keep track of the sequence.

**Reply by dvarrazzo** [reply]
Why would setting the default as a temporary sequence be such a bad idea? You can add the sequence/default, copy data in, remove the default, in the same transaction.

**Reply by greyltc** [reply]
I actually had that exact thing in testing for a bit. The table can be concurrently written to from other places and I got worried about collisions on changes to the default setting on the column from various other writers setting the default to their own temporary sequences. Do you think that's a valid concern?

**Reply by dvarrazzo** [reply]
If you do all in a transaction, because altering a table takes an exclusive lock, a concurrent process would be blocked until the previous one has finished.

If any, I wonder where do you start the sequence: from a "SELECT max(id)“? To make sure to maintain consistency it's probably better to take an exclusive lock on the table.

Working this all in a transaction has the advantage of simplicity, but operations are serialised. If you eant to run parallel operations you need some init/cleanup process managing the alter table outside a transaction, but coordination is more difficult.
